# Exception Handling

에러가 발생하면 수행하던 작업은 전부 수행 전 상태로 되돌려야 한다.

메서드가 객체의 상태를 바꾸는 작업을 한다면, 예외가 발생한 경우 객체의 상태를 원 상태로 되돌리는 것이 원칙이다. 이것을 실패 원자적(failure-atomic) 이라고 한다.

100건 수행하는 메서드가 있을 때

```javascript
try {
    for(1~100) {
        executeOne(); 한 건씩 수행, 실패 시 error
    }
} catch(err) {
    1. 에러페이지가 필요한 경우 에러페이지
    2. log 기록이 필요한 경우 logger
    3. 100건에 대하여 전부 수행 전 상태로 되돌린다
}

executeOne() {
    try {
        logic();
    } catch(err) {
        throw new Error(err);
    }
}

logic() {
    try {
        function1(); 
        function2();
    } catch(err) {
        Logger.error(err);
        throw new Error(err)
    }
}
```

의도에 따라 다르겠지만, 예외 처리가 불가능한 계층에서의 예외는 상위 계층으로 던지고, 예외 처리가 가능한 계층에서 예외를 처리해야한다.

위에서는 function 계층이나 logic 계층에서 예외 처리가 불가능하기 때문에 상위 계층으로 예외를 전달했다.

여기서 한 건마다 예외를 처리할 것인지, 100건 전부의 예외를 한번에 처리할 것인지에 따라 달라진다. 위에서는 100건 전부의 예외를 한번에 처리했다.

### 메서드를 실패 원자적으로 만드는 방법

1. 작업 수행에 앞서 매개변수의 유효성을 검사한다. (typeof 검사, size 검사 등)
2. 객체의 임시 복사본에서 작업을 수행한 다음, 작업이 성공하면 원래 객체와 교체한다.
